Migration to ReadTheDocs

JonathanWenger opened the pull request:
This PR migrates the documentation from being built by Travis and being deployed to the `gh-pages` branch to being built and hosted by ReadTheDocs. This has the primary advantage that different versions of the documentation for different branches can be easily built and viewed in parallel.

**Note:** This changes the domain of the documentation from https://probabilistic-numerics.github.io/probnum/ to https://probnum.readthedocs.io

**Todos:**
- [ ] merge PR into `development`
- [ ] merge `development` into `master`
- [ ] Update settings in readthedocs dashboard
- [ ] Update repository settings to not serve the `gh-pages` branch
- [ ] Delete `gh-pages` branch
- [ ] ~remove travis's github token~

JonathanWenger commented:
This closes #84 and #7 .

JonathanWenger commented:
You can view the documentation generated by this PR at https://probnum.readthedocs.io. As soon as the PR is merged I will swap the underlying repo at ReadTheDocs to `probabilistic-numerics/probnum` from `JonathanWenger/probnum`.

JonathanWenger commented:
**Note**: Docs for master will not build at `probnum.readthedocs.io` until changes from this PR are merged into master.

nathanaelbosch commented on CONTRIBUTING.md:
```diff
@@ -69,23 +69,30 @@ To run the test suite on your machine you have multiple options:
 
 - **Full test suite with tox:** Run the full suite across different Python versions with
   
```
Personally, I would prefer either not having empty lines in this list before the code blocks, or having empty lines both before and after.

JonathanWenger replied:
No empty lines breaks the conversion to html => empty lines before and after added.

nathanaelbosch commented on CONTRIBUTING.md:
```diff
@@ -69,23 +69,30 @@ To run the test suite on your machine you have multiple options:
+[Travis](https://travis-ci.org/probabilistic-numerics/probnum) and hosted by [ReadTheDocs](https://readthedocs.org/projects/probnum/).
+You can access the documentation for the default version and the development branch here:
+
+[![default docs](https://img.shields.io/readthedocs/probnum.svg?logo=read%20the%20docs&logoColor=white&label=Docs)](https://probnum.readthedocs.io)
```
I like "stable" and "latest" to refer to the two different states, but "default" and "development" is of course also fine for me. I don't really know how others call those.

JonathanWenger replied:
In the latest and stable language latest would be `master` and `stable` a release. Is this what we want at this point? This would also not expose the developer docs. My suggestion would be to have `stable`:release, `latest`:master, `dev`:development

nathanaelbosch replied:
Then I might have misunderstood the meaning of stable/latest. The cases in which I saw that naming seemed to follow a workflow in which they work directly on master, without using a development branch. Therefore I assumed that the mapping to our workflow would be to use stable/latest for master/development, but this might very well be completely false. 

I like your proposal of using:
- stable:release
- latest:master
- dev:development

nathanaelbosch commented on README.md:
```diff
@@ -1,7 +1,8 @@
 # Probabilistic Numerics in Python
 
-[![Build Status](https://travis-ci.org/probabilistic-numerics/probnum.svg?branch=master)](https://travis-ci.org/probabilistic-numerics/probnum)
+[![Build Status](https://img.shields.io/travis/probabilistic-numerics/probnum/master?logo=travis%20ci&logoColor=white&label=Travis%20CI)](https://travis-ci.org/probabilistic-numerics/probnum)
```
As I understood it, the banner from travis-ci.org shows the actual build status. Personally I would prefer that and consider it to be a bit more informative than an "always green" banner

JonathanWenger replied:
This should still be the behaviour. Compare also: https://github.com/astropy/astropy/blob/master/README.rst

nathanaelbosch replied:
My bad, I completely misjudged the capabilities of shields.io!

nathanaelbosch commented on README.md:
```diff
@@ -1,7 +1,8 @@
-[![Build Status](https://travis-ci.org/probabilistic-numerics/probnum.svg?branch=master)](https://travis-ci.org/probabilistic-numerics/probnum)
+[![Build Status](https://img.shields.io/travis/probabilistic-numerics/probnum/master?logo=travis%20ci&logoColor=white&label=Travis%20CI)](https://travis-ci.org/probabilistic-numerics/probnum)
 [![Coverage Status](http://codecov.io/github/probabilistic-numerics/probnum/coverage.svg?branch=master)](http://codecov.io/github/probabilistic-numerics/probnum?branch=master)
+[![Documentation](https://img.shields.io/readthedocs/probnum.svg?logo=read%20the%20docs&logoColor=white&label=Docs)](https://probnum.readthedocs.io)
```
Should we have two banners, one for the "default" docs and one for "development"? (or "stable" and "latest", or whatever we decide to use)

JonathanWenger replied:
In light of a merger into `master` in the near future, I would vote for simplicity here and just have a link to `latest` (i.e. `master`). This also seems to be the standard in similar repos.

nathanaelbosch replied:
Sounds good :)

nathanaelbosch commented on setup.cfg:
```diff
@@ -19,7 +19,7 @@ classifiers =
     License :: OSI Approved :: MIT License
     Operating System :: OS Independent
-    Programming Language :: Python :: 3.5
+    Programming Language :: Python :: 3.6
```
Nice catch, never realized that we had mismatched versions in this specification and the testing!

nathanaelbosch commented on .readthedocs.yml:
```diff
@@ -0,0 +1,23 @@
+
+# Build documentation in the docs/ directory with Sphinx
+sphinx:
+  builder: html
```
I assume this means that the documentation is now not actually built through tox anymore?
If yes, we could adjust tox to use the same (or a very closely matching) command.

nathanaelbosch replied:
Or I guess that would not really change much, so I'm fine with leaving it the way it is.

JonathanWenger replied:
Judging from other repos having the same setup there is no way to have read the docs actually run the tox command. Instead it just directly calls sphinx-build.

nathanaelbosch commented on .travis.yml:
```diff
@@ -12,10 +12,4 @@ install:
-  local_dir: docs/_build/html
-  github_token: $GITHUB_TOKEN # Set in travis-ci.org dashboard
-  keep_history: true
+  - bash <(curl -s https://codecov.io/bash) # upload test coverage to codecov
```
**Really unimportant and optional:** If you end up making any changes, you could add a newline here.

nathanaelbosch commented on .readthedocs.yml:
```diff
@@ -0,0 +1,23 @@
+  install:
+    - method: pip
+      path: .
+    - requirements: docs/requirements.txt
```
**Really unimportant and optional:** If you end up making any changes, you could add a newline here.

nathanaelbosch commented:
I also wondered: In this PR you ported the documentation in its current style to readthedocs. Does this only represent a migration to another domain (with additional functionality regarding multiple documentations of course!), or does this also include a statement/decision about the stylistic choice for the ProbNum documentation, i.e. "our current layout" vs "the common readthedocs layout"?

JonathanWenger commented:
> I also wondered: In this PR you ported the documentation in its current style to readthedocs. Does this only represent a migration to another domain (with additional functionality regarding multiple documentations of course!), or does this also include a statement/decision about the stylistic choice for the ProbNum documentation, i.e. "our current layout" vs "the common readthedocs layout"?

In order to reach a decision I ported our thematic choices of the docs to the readthedocs template. See here: https://probnum.readthedocs.io/en/docs-style/index.html 
We might want a third opinion (e.g. @pnkraemer) here on what is better. Note that the documentation structure in the readthedocs template is still subject to change (and be more coherent).

nathanaelbosch commented:
Awesome work Jonathan! I'm really impressed by seeing the documentation in both styles.

Personally, I think I slightly prefer the readthedocs layout, but I find it really hard to decide between both. But it really is awesome that we can compare both side by side :)

JonathanWenger commented:
This PR is ready to merge but kept in draft mode since it breaks the documentation links on `master`. This can be merged into `development` just prior to a merge of `development` into master.

pnkraemer commented:
>  We might want a third opinion (e.g. @pnkraemer) here on what is better. Note that the documentation structure in the readthedocs template is still subject to change (in particular the structure could be more coherent).
Sorry for getting back to you so late, seems like it got lost somehow. In case my opinion is still wanted: I prefer the readthedocs style if the website is _just_ for documentation of the code. If it is supposed to serve as some kind of website for PN related things (not sure exactly what I am thinking of but lets be imaginative here), then the previous style was nicer.

JonathanWenger commented:
**Update**:
The following larger changes are made by this commit:
- The documentation will from now on be hosted by ReadTheDocs at https://probnum.readthedocs.org
- The documentation can be viewed for different branches and releases
- The documentation style now is based off of `sphinx_rtd_theme` instead of bootstrap

During this migration I also changed much of the structure of the documentation and added some new sections to the documentation. During review I am mostly interested in feedback on the following:
**General:**
- [ ] Menu structure
- [ ] Intuitiveness of page (organization, page names, is information easy to find?)
**Added Pages:**
- [Index Page](https://probnum.readthedocs.io/en/docs-update/index.html)
- Introduction
  - [Probabilistic Numerics](https://probnum.readthedocs.io/en/docs-update/introduction/probabilistic_numerics.html)
  - [ProbNum Quickstart](https://probnum.readthedocs.io/en/docs-update/introduction/quickstart.html)

nathanaelbosch commented on .readthedocs.yml:
```diff
@@ -0,0 +1,23 @@
+  install:
+    - method: pip
+      path: .
+    - requirements: docs/requirements.txt
```
This means we have some abiguity regarding "building the docs" now, right? What I mean is that tox does not perfectly mirror the way the docs are built anymore.

JonathanWenger replied:
I wasnt able to figure out a way to have ReadTheDocs use tox. ReadTheDocs also doesnt seem to support using a custom Makefile to build the docs (see also: https://github.com/readthedocs/readthedocs.org/issues/2423). I think the closest we can get for reproducibility is to specify a common `docs/requirements.txt`. ReadTheDocs then calls `sphinx-build` like the makefile, but potentially with different options which might cause divergent behaviour, I guess.

nathanaelbosch replied:
I think it should be fine this way :) Maybe one thing we could do is to have tox use the same `./docs/requirements.txt` file

nathanaelbosch replied:
Okay I just realized that you already adjusted `tox.ini`, awesome!

nathanaelbosch commented on docs/source/development/documentation.rst:
```diff
@@ -1,3 +1,127 @@
+**General Rules**
+
+- Cover :code:`Parameters`, :code:`Returns`, :code:`Raises` and :code:`Examples` at every publicly visible docstring---in that order.
+- Examples are tested via doctest. Ensure :code:`doctest` does not fail.
```
Do we actually run doctests right now? Is that included in pytest by default?

JonathanWenger replied:
In the `setup.cfg` file there is an option configuring `pytest`. I also added `--doc-test` to the `tox.ini` now. This generates this output indicating that doctests are run:
```bash
src/probnum/diffeq/ode/ivp.py::src.probnum.diffeq.ode.ivp.IVP PASSED     [  0%]
src/probnum/diffeq/odefiltsmooth/odefiltsmooth.py::src.probnum.diffeq.odefiltsmooth.odefiltsmooth.probsolve_ivp PASSED [  0%]
src/probnum/linalg/linearsolvers/linearsolvers.py::src.probnum.linalg.linearsolvers.linearsolvers.problinsolve PASSED [  1%]
src/probnum/linalg/linops/linearoperators.py::src.probnum.linalg.linops.linearoperators.LinearOperator PASSED [  1%]
src/probnum/linalg/linops/linearoperators.py::src.probnum.linalg.linops.linearoperators.aslinop PASSED [  2%]
src/probnum/prob/randomvariable.py::src.probnum.prob.randomvariable.asrandvar PASSED [  2%]
src/probnum/prob/distributions/dirac.py::src.probnum.prob.distributions.dirac.Dirac PASSED [  2%]
src/probnum/prob/distributions/normal.py::src.probnum.prob.distributions.normal.Normal PASSED [  3%]
src/probnum/quad/polynomial/clenshawcurtis.py::src.probnum.quad.polynomial.clenshawcurtis.ClenshawCurtis PASSED [  3%]

```

nathanaelbosch replied:
Cool, thanks!

nathanaelbosch commented on docs/source/development/documentation.rst:
```diff
@@ -1,3 +1,127 @@
+	This package implements common operations and (probabilistic) numerical methods for linear algebra.
+	"""
+
+	from probnum.linalg.linearsolvers import *
```
I guess this concludes that we indeed do not want to check for PEP-8 errors regarding star imports in `__ini__.py` files

JonathanWenger replied:
I forgot to change this. Based on the result of our discussion explicit imports need to be specified in the `__all__` variable anyway, therefore making explicit imports little additional work.

nathanaelbosch replied:
I agree that in general we should avoid star imports. But for some reason when I looked into fixing those issues in `__init__.py` files I came to the conclusion that there might also be reasons for having an exception to this rule in those files. 

Anyways, thanks for fixing it in the documentation, this way at least we don't explicitly encourage contributors to use them :)

nathanaelbosch commented on docs/source/index.rst:
```diff
@@ -1,21 +1,91 @@
+
+.. code-block:: shell
+   
+   pip install probnum
```
Is this valid already? Or will this be valid as soon as we push to master?

JonathanWenger replied:
Not actually valid yet. My goal was to resolve this asap after the new docs. This does create a period of the documentation being inaccurate though.

nathanaelbosch replied:
Thanks, good to know!

nathanaelbosch commented on docs/source/introduction/probabilistic_numerics.rst:
```diff
@@ -0,0 +1,39 @@
+
+`Probabilistic numerics (PN) <http://probabilistic-numerics.org/>`_ aims to quantify uncertainty arising from intractable or incomplete numerical computation and from stochastic input. This new paradigm which has emerged at the intersection of computer science and numerical analysis treats a *numerical problem* as one of *statistical inference* instead. [#]_ [#]_ 
+
+The probabilistic viewpoint provides a principled way to encode structural knowledge about a problem. By giving an explicit role to uncertainty from all sources, in particular from the computation itself, PN gives rise to new applications beyond the scope of classical methods. For example, it makes consistent propagation of uncertainty the entirety of a computational pipeline possible.
```
Typo in last sentence: "it makes [...] **over** the entirety of a computational pipeline possible"

nathanaelbosch commented:
My feedback, as discussed:
- [ ] use api names for the api documentation: "`probnum.diffeq`" instead of "Differential Equations", "`probnum.utils`" instead of "Other", etc
- [ ] different tutorial structure:
  - Remove "Case Studies" and "Advanced" since they are empty
  - Maybe use that level of structure (where "Advanced" was) to structure into diffeq, linalg, etc
